File: Leetcode/CountNumberofMaximumBitwise-ORSubsets.rs
// https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets

struct Solution {}

impl Solution {
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let mut cnt = vec![0; 1 << 18];

        let mut ans = 0;
        for mask in 0..(1 << nums.len()) {
            let mut s = 0;
            for b in 0..16 {
                if mask & (1 << b) > 0 {
                    s |= nums[b];
                }
            }
            cnt[s as usize] += 1;
            ans = ans.max(cnt[s as usize]);
        }

        ans
    }
}

File: Leetcode/CountNumberofMaximumBitwise-ORSubsets.cpp
// https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets

class Solution {
   public:
    int countMaxOrSubsets(vector<int>& nums) {
        const auto n = nums.size();

        auto ma = 0;
        for (const auto& num : nums) {
            ma |= num;
        }

        auto ans = 0;
        for (auto mask = 0; mask < (1 << n); ++mask) {
            auto cur = 0;
            for (auto bit = 0; bit < n; ++bit) {
                if (mask & (1 << bit)) {
                    cur |= nums[bit];
                }
            }
            ans += (cur == ma);
        }

        return ans;
    }
};